content/blog/elm-at-noredink/index.md

```diff
@@ -0,0 +1,445 @@
+        "teach-classes-elm"
+    )
+```
+
+The `NriProgram` that you see here is a small wrapper around common
+operations that we need to perform, such as:
+- passing an environment to our Elm apps so that we can have different
+  settings in test, development and production
+- setting up some basic analytics and reporting
+- grabbing the flags and the div where the Elm app will be mounted
+- giving us the ability to easily set up ports using `subscribe` and `send`
+
+Another interesting bit is the convention that we use to name our Elm
+modules. Since the URL is going to be `/teach/classes`, the associated Elm
+module is going to be `Page.Teach.Classes.Main`. Here are some other
+examples:
+
+- `Page.Admin.RelevantTerms.Main` corresponds to `/admin/relevant_terms`
+- `Page.Learn.ChooseSourceMaterials.Main` corresponds to `/learn/choose_source_materials`
+- `Page.Preferences.Main` corresponds to `/preferences`
+- `Page.Teacher.Courses.Assignments.Main` corresponds to `/teach/courses/:id/assignments`
+
+In total, we have around 80 Elm apps that serve different Rails
+controllers. They consist of a mixture of normal applications and single page
+applications. By adapting the Rails motto "convention over configuration"
+we're confident we can scale this approach indefinitely. Let's take a deeper look!
+
+## Our Elm apps
+
+At this point, you probably won't be surprised to learn that we use a
+custom wrapper as the entrypoint of our Elm programs. Here is how it looks
+like:
+
+```elm
+main : Nri.Program.Program Model Msg
+main =
+    Nri.Program.program
+        { moduleName = "Page.Interests.Main"
+        , flagsDecoder = \_ -> decode
+        , perform = perform
+        , init = init
+        , view = view
+        , update = update
+        , subscriptions = \_ -> Sub.none
+        }
+```
+
+Let's compare it to `Browser.element` from `elm/browser`:
+
+```elm
+main : Browser.Program flags model msg
+main =
+    Browser.element
+        { init = init
+        , view = view
+        , update = update
+        , subscriptions = \_ -> Sub.none
+        }
+```
+
+As you can see, the types are slightly different and we include a couple of
+different fields in our version.
+
+What are the advantages of using a custom `Nri.Program`?
+
+- reducing boilerplate: for example, we can automatically handle decoding failures so that
+  we don't have to worry about them in each individual page.
+- handling generic behaviours: for example, we can automatically detect the user's
+  input method to make our product more accessible.
+
+The differences don't stop there. Let's look at the type of the `view` function:
+
+```elm
+view : Env -> Model -> Html Msg
+```
+
+Here we take an additional `Env` argument. This record contains information
+about the current release, the logged in user and the environment where the
+code is running. For example, if we detect that we are running in test
+mode, there is no need to run fancy UI animations, thus we can disable them.
```

Comment:
🎨 feel free to skip, but mentioning that disabling animations also makes our integration tests more reliable might be an interesting point. otherwise, I would maybe lean towards leaving them on to make sure tests run the exact same code as prod when possible.